### PR TITLE
`rustc_data_structures` cleanups

### PR DESCRIPTION
**Pull request by nnethercote**
Some improvements I found while looking through this code.

r? @michaelwoerister

### CODE REVIEW

File: compiler/rustc_middle/src/mir/interpret/mod.rs [hunk 1]
```diff
@@ -149,15 +131,28 @@ pub use self::error::{
...
+use rustc_macros::{HashStable, TyDecodable, TyEncodable, TypeFoldable, TypeVisitable};
+use rustc_middle::ty::print::with_no_trimmed_paths;
+use rustc_serialize::{Decodable, Encodable};
+use rustc_target::abi::{AddressSpace, Endian, HasDataLayout};
+use smallvec::{smallvec, SmallVec};
+use std::fmt;
+use std::io;
+use std::io::{Read, Write};
+use std::num::NonZero;
+use std::sync::atomic::{AtomicU32, Ordering};
```
Comment (RalfJung) [inline compiler/rustc_middle/src/mir/interpret/mod.rs:155]: I would prefer if you did not reorder all the imports here.

if you must reorder them, please follow the style of: std, non-rustc dependencies, rustc dependencies, current crate.

File: compiler/rustc_middle/src/mir/interpret/mod.rs [hunk 2]
```diff
@@ -357,7 +350,7 @@ impl<'s> AllocDecodingSession<'s> {
                         bug!("this should be unreachable");
                     } else {
                         // Start decoding concurrently.
-                        sessions.insert(self.session_id);
+                        sessions.insert(0, self.session_id);
```
Comment (RalfJung) [inline compiler/rustc_middle/src/mir/interpret/mod.rs:353]: This changes an O(1) operation to an O(n) operation. Should we be concerned?

(I don't know anything about this code, I am just looking at the diff)

**Review by michaelwoerister — COMMENTED** [review]
Thanks for the PR, @nnethercote! Looks good to me overall.

Let's do a perf run to see if there is a difference due to the case @RalfJung points out.

File: compiler/rustc_hir/src/lang_items.rs [hunk 3]
```diff
@@ -55,17 +55,22 @@ macro_rules! language_item_table {
...
-                    ///
-                    $(#[$attr])*
-                    $variant,
-                )*
+        impl LangItem {
+            pub fn from_u32(u: u32) -> Option<LangItem> {
+                $(if u == LangItem::$variant as u32 {
+                    return Some(LangItem::$variant)
+                })*
+                None
```
Comment (SparrowLii) [inline compiler/rustc_hir/src/lang_items.rs:73]: This function is only used in the `LanguageItems::iter()`. And each iteration will traverse the whole `LangItem`, which is not efficient. Maybe use a static array to iterate instead?:
```
impl LangItem {
    pub const fn array() -> [LangItem; std::mem::variant_count::<LangItem>()] {
        [$(LangItem::$variant,)*]
    }
}
```

File: compiler/rustc_middle/src/mir/interpret/mod.rs (hunk 1 above)
Comment (tgross35) [reply compiler/rustc_middle/src/mir/interpret/mod.rs:155]: Could we just set rustfmt [`group_imports`](https://rust-lang.github.io/rustfmt/?version=v1.6.0&search=#group_imports) to `StdExternalCrate` somewhere?

Comment (RalfJung) [reply compiler/rustc_middle/src/mir/interpret/mod.rs:155]: Maybe... though that will mix rustc_ crate imports with fully external imports, it seems? Or can rustfmt take into account which crates are part of the local workspace?

Comment (RalfJung) [reply compiler/rustc_middle/src/mir/interpret/mod.rs:155]: Also https://github.com/rust-lang/rustfmt/issues/4709 makes this a complete no-go I think.

Comment (nnethercote) [reply compiler/rustc_middle/src/mir/interpret/mod.rs:155]: > If you must reorder them, please follow the style of: std, non-rustc dependencies, rustc dependencies, current crate.

I estimate that approximately zero files in the codebase use that ordering. There's unfortunately not a lot of consistency about how `use` items are sectioned and ordered, though "put them all in one alphabetical section" is dead simple, possibly the most common, how I usually do it, and nobody has ever complained about it before.

File: compiler/rustc_hir/src/lang_items.rs (hunk 3 above)
Comment (nnethercote) [reply compiler/rustc_hir/src/lang_items.rs:73]: I am preserving the existing behaviour. I doubt this code is perf-sensitive?

File: compiler/rustc_middle/src/mir/interpret/mod.rs (hunk 1 above)
Comment (RalfJung) [reply compiler/rustc_middle/src/mir/interpret/mod.rs:155]: I use that ordering in files I maintain.

I ask you not mess with the order of other files when editing parts of the compiler for which you are not the primary maintainer.

File: compiler/rustc_hir/src/lang_items.rs (hunk 3 above)
Comment (SparrowLii) [reply compiler/rustc_hir/src/lang_items.rs:73]: That's OK

File: compiler/rustc_middle/src/mir/interpret/mod.rs (hunk 1 above)
Comment (RalfJung) [reply compiler/rustc_middle/src/mir/interpret/mod.rs:155]: > I estimate that approximately zero files in the codebase use that ordering.

Counterexamples: [1](https://github.com/rust-lang/rust/blob/875f0c2da05da9a7620afd6e2be04fbcb9a4b395/compiler/rustc_const_eval/src/interpret/place.rs#L5-L14), [2](https://github.com/rust-lang/rust/blob/4497990dff8ce181fa12757eb4e786b85025ae61/compiler/rustc_const_eval/src/interpret/projection.rs#L10-L20), [3](https://github.com/rust-lang/rust/blob/bb8d6f790b63797a816a4007e8ab002ed2906de7/compiler/rustc_const_eval/src/interpret/operand.rs#L4-L14), [4](https://github.com/rust-lang/rust/blob/ae7b07f2dc9b07aceab318829538222666945efc/compiler/rustc_const_eval/src/interpret/memory.rs#L9-L29), [5](https://github.com/rust-lang/rust/blob/126dcc618d065e97887113926f63986bd0a9e681/compiler/rustc_const_eval/src/interpret/intern.rs#L16-L29).

I think about 80% of the files in the `interpret` folder follow that ordering (and some of the ones that don't still almost do, they may just have a `crate` import separate from and *after* a `self` import which is a bit odd).

File: compiler/rustc_hir/src/lang_items.rs (hunk 3 above)
Comment (nnethercote) [reply compiler/rustc_hir/src/lang_items.rs:73]: On second thoughts, this is a good suggestion. I will do it.

File: compiler/rustc_middle/src/mir/interpret/mod.rs (hunk 1 above)
Comment (nnethercote) [reply compiler/rustc_middle/src/mir/interpret/mod.rs]: I have reverted that part of the change. That file is strange, with macros and `mod` items before the `use` items.

It is unfortunate that rustfmt is so unopinionated about `use` item ordering, and that `group_imports` is not mature enough to be usable.

> I ask you not apply your own style for the order of imports when editing parts of the compiler for which you are not the primary maintainer.

I'm going to push back strongly here. I have no problem with people arguing against changes on the merits, but this sentence goes well beyond that. "Primary maintainer" is not something that has any official meaning within the Rust project. Nor does the project have any principle that suggests a commit X would be acceptable if done by person A but unacceptable if done by person B. Please don't invent process in an attempt to block a change.

**Comment by nnethercote** [comment]
@michaelwoerister I have addressed the comments. I don't think a perf run is necessary now that I changed the `insert(0, ...)` calls to `push()`.

File: compiler/rustc_middle/src/mir/interpret/mod.rs (hunk 1 above)
Comment (RalfJung) [reply compiler/rustc_middle/src/mir/interpret/mod.rs]: I should have worded this better.  We don't have a formal concept of "primary maintainer", agreed, but we have de-facto people focusing on particular parts of the compiler, and being responsible for design decisions affecting that part. Larger interpreter changes will involve wg-const-eval. Formatting changes are not quite the same, but we also do have a policy against formatting-only changes, except when accompanied by tool support (and, where appropriate, backed by team consensus). This was established in reply to people making sweeping PRs that did nothing but add trailing `.` to comments or things like that. I don't know where we document such policy decisions.

I agree the file is strange, and I wouldn't be opposed to some sort of reordering that doesn't change the formatting of `use` -- TBH I never found a good style for how to order `mod` and `use`. IIRC the macros have to (or had to) go first because of macro magic; not sure if that still applies.

Thanks for reverting the changes.

File: compiler/rustc_middle/src/mir/interpret/mod.rs [hunk 4]
```diff
@@ -125,10 +125,12 @@ use std::io::{Read, Write};
 use std::num::NonZero;
 use std::sync::atomic::{AtomicU32, Ordering};
 
+use smallvec::{smallvec, SmallVec};
+use tracing::debug;
```
Comment (RalfJung) [inline compiler/rustc_middle/src/mir/interpret/mod.rs:129]: Why is this import needed now? The diff doesn't otherwise mention `debug!`...

Comment (nnethercote) [reply compiler/rustc_middle/src/mir/interpret/mod.rs]: It was a rebasing mix-up relating to the third commit removing `extern crate tracing`. I have removed it.

Comment (nnethercote) [reply compiler/rustc_middle/src/mir/interpret/mod.rs]: Because `debug!` is used, *and* there is a `#[macro_use] extern crate tracing`, so there are no warnings/errors whether the `use` is present or not.

File: compiler/rustc_hir/src/lang_items.rs [hunk 5]
```diff
@@ -55,21 +52,21 @@ macro_rules! language_item_table {
...
+            $(
+                #[doc = concat!("The `", stringify!($name), "` lang item.")]
+                $(#[$attr])*
+                $variant,
+            )*
         }
 
         impl LangItem {
+            const fn items() -> [LangItem; std::mem::variant_count::<LangItem>()] {
+                [$(LangItem::$variant,)*]
```
Comment (michaelwoerister) [inline compiler/rustc_hir/src/lang_items.rs:67]: Can we add an assertion somewhere that `LangItem::items()[lang_item as usize] == lang_item`. Just in case something changes with discriminant assignment.

**Review by michaelwoerister — COMMENTED** [review]
One last comment. With that assertion added, you can r=me.

File: compiler/rustc_hir/src/lang_items.rs (hunk 5 above)
Comment (nnethercote) [reply compiler/rustc_hir/src/lang_items.rs]: I went to do that, and then I wrote a comment explaining the assumption, and then I decided to revert to the original branch-based function, because eliminating the macro is the main point of this change. Someone can do a follow-up if they like to do the array thing. I did add a comment explaining that the branch-based function doesn't make any assumptions about discriminant ordering.

Comment (michaelwoerister) [reply compiler/rustc_hir/src/lang_items.rs]: That works for me.
